### ISSUE TITLE
Evaluate alternative transaction fee market mechanisms

### ISSUE THREAD

**Issue by vbuterin** [zcash/zcash#3473]
I've presented the ideas here online and at multiple events including most recently CRYPTO 2018 today, and I got a positive reception from some Zcash community members and it was suggested that I start a discussion on this topic, so here goes.

Currently, most public blockchains, including Ethereum, Bitcoin, and (as I understand) Zcash, use a block size limit and a fee market to regulate the inclusion of transactions and prevent users from spamming the blockchain, forcing transaction senders to pay for the costs they impose on the network. However, this cap-and-auction approach is only one of many ways of achieving thhis goal, and there are other mechanisms that are worth exploring, that I think are better for several reasons.

There are three major problems with the status quo of transaction fee markets:

* **Mismatch between volatility of transaction fee levels and social cost of transactions**: transaction fees on mature public blockchains, that have enough usage so that blocks are full, tend to be extremely volatile. On Ethereum, minimum fees are typically around 2 gwei (10^9 gwei = 1 ETH), but sometimes go up to 20-50 gwei and have even on one occasion gone up to over 200 gwei: https://etherscan.io/chart/gasprice. This clearly creates many inefficiencies, because it's absurd to suggest that the cost incurred by the network from accepting one more transaction into a block actually is 100x more when gas prices are 200 gwei than when they are 2 gwei; in both cases, it's a difference between 8 million gas and 8.02 million gas.
* **Inefficiencies of first price auctions**: see https://ethresear.ch/t/first-and-second-price-auctions-and-improved-transaction-fee-markets/2410 for a detailed writeup. In short, the current approach, where transaction senders publish a transaction with a fee, miners choose the highest-paying transactions, and everyone pays what they bid, is well-known in mechanism design literature to be highly inefficient, and so complex fee estimation algorithms are required, and even these algorithms often end up not working very well, leading to frequent fee overpayment. See also https://blog.bitgo.com/the-challenges-of-bitcoin-transaction-fee-estimation-e47a64a61c72 for a Bitcoin core developer's description of the challenges involved in fee estimation in the status quo.
* **Instability of blockchains with no block reward**: in the long run, blockchains where there is no issuance (including Bitcoin and Zcash) at present intend to switch to rewarding miners entirely through transaction fees. However, there are [known results](http://randomwalker.info/publications/mining_CCS.pdf) showing that this likely leads to a lot of instability, incentivizing mining "sister blocks" that steal transaction fees, opening up much stronger selfish mining attack vectors, and more. There is at present no good mitigation for this.

I will suggest an alternative mechanism that mitigates all three of the issues. The protocol internally maintains a fee level `f`, and a miner that creates a block that includes `d` bytes (or gas, or weight units, or whatever) must pay a fee of `f * d`, which gets put into a pot. There is a maximum weight limit `M`. The protocol adjusts `f` so that long-run block space usage averages out to `M/2`; it can do this for example by adjusting `f[n+1] = f[n] * (1 + 0.2 * w[n] / M)`, where `f[n]` is the fee level in the current block, `f[n+1]` is the fee level for the next block, and `w[n]` is the total bytes/gas/weight consumed in the current block. In every block, a miner gets a reward equal to 1/N (eg. 1/10000) of the money remaining in the pot (note that this amount does NOT depend on the transactions they include in their block).

This accomplishes the following goals:

* It mitigates the economic inefficiencies from social cost mismatch due to fee volatility. There is a fairly nuanced economic argument here; see particularly pages 16-20 of the paper linked in https://ethresear.ch/t/draft-position-paper-on-resource-pricing/2838 (though I recommend reading the whole paper) for a detailed argument of why this is the case. Intuitively, the adjusting fee mechanism works like a fixed fee in the short run and a cap in the long run, and it turns out that because of [arguments from Martin Weitzman's 1974 paper](https://scholar.harvard.edu/weitzman/files/prices_vs_quantities.pdf) fixed fees are likely better than a cap in the circumstances in which we're actually presented in public blockchains.
* It replaces the auction with a fixed price sale (except during short periods where blocks fill up completely until fees catch up), making fee estimation extremely simple: calculate the fee `f` for the next block, if you can afford it pay it, otherwise don't. This eliminates inefficiencies from the first price auction mechanism.
* It creates a mechanism similar to a permanent block reward (the 1/N coming from the pot), mitigating many of the instability issues with fee-only blockchains without requiring actual permanent issuance.

This is something that's being discussed in the context of Ethereum but is theoretically relevant to any public blockchain, so is certainly something that the Zcash community could consider.

**Comment by ch4ot1c** [comment]
Exciting research - incredible work as always. I think your proposals and paper improve the model that all chains today use, socially speaking. Balanced/fair costs between transactors and miners. A smoother ride for all.

I don't think it makes a lot of sense to transition a chain from an 'auction' fee-selection system to a sliding one - it could cause a contentious fork / more trouble than good. 

Fee estimation algorithms being 'hard' is sorta a scapegoat, since they don't incur a heavy time cost (only once every 10 / 2.5 minutes, only for the miner, for all miners). If the transactors are the ones experiencing the cost inefficiency, then they just need to bid smarter (low fee, and wait a while).

> This clearly creates many inefficiencies, because it's absurd to suggest that the cost incurred by the network from accepting one more transaction into a block actually is 100x more when gas prices are 200 gwei than when they are 2 gwei; in both cases, it's a difference between 8 million gas and 8.02 million gas.

I'm unconvinced that this is an inefficiency and not an effect; the condition of whether an additional tx is allowed is all-or-none (whether it fits), for-the-next-block. The selection of 'best' via maximizing the fees is a memoryless, miner-only activity. For an equal sized transaction, under different user weather, indeed the mining conditions are that much different. It is not as much an outrageous cost to the spender, as it is a freebie for the miner. The network never has any 'costs', besides its (free) intermittent coinbase payouts; its integrity would be destroyed though if it were to allow that marginal transaction.

The improvements to the social economics look worthwhile, but probably only work in practice on a new blockchain. It'd be a speculative play to hardfork with the post-21m or volatility conditions in mind.

I wonder what the Zcash team thinks about all this  🤔 

Thank you for the citations!

**Comment by vbuterin** [comment]
> Fee estimation algorithms being 'hard' is sorta a scapegoat, since they don't incur a heavy time cost (only once every 10 / 2.5 minutes, only for the miner, for all miners)

Fee estimation algorithms are client-side, not miner-side. So all wallet implementations need to have them, and if/once proprietary implementations start getting developed that are reliably better, that's a possible centralization vector.

>  The network never has any 'costs'

Ah, by "costs" I was referring to costs incurred by the participants and beneficiaries of the network, including cost of processing, cost of storage, and marginal increments to systemic risks involving network centralization. So not the limit itself, but the factors that led to the limit being instituted in the first place. The issue is that the risks don't actually increase much if the blockchain sometimes has 7m gas per block and sometimes 9m, versus always having 8m gas per block, and a flexible schedule that allows usage to drop to 7m when fees are low and rise to 9m when fees are high make the blockchain more usable by allowing more people to participate when demand is highest.

>  it could cause a contentious fork / more trouble than good ... the improvements to the social economics look worthwhile, but probably only work in practice on a new blockchain

Oh sure. If the Zcash community values stability of protocol, or not entering into weird unknown unknowns, enough that this is too contentious, then yes that's a reason not to adopt it. That said, from my observations I don't feel like the Zcash community is _that_ conservative; it's willing to adopt radical changes to its underlying cryptography with a less than year-long development cycle, use weird spooky crypto with trusted setups, etc. I feel that the one thing that Zcash should definitely absolutely not take risks on is its core value proposition, which is giving people privacy. Possibly the 21m limit as well if that's the community's wish. I think transaction fee economics is more like a technical feature that can and should be optimized, and not some fundamental immutable feature that must be safeguarded in its current form at all costs. That said, it's ultimately up to Zcash users to decide this.

(a moderate third alternative is of course to sit for two years and let ethereum be the guinea pig for changes to transaction fee economics first)

**Comment by DarkEmi** [comment]
Hello Vitalik,

Funnily I submitted a very similar idea 2 years ago (with some variations but its a variation on your proposal) here : 
https://medium.com/@lespectrenoir42/bitcoin-proposal-delayed-fee-market-6d660cfd9793

It had the same goals, mainly suppressing the inneficient auction mechanism and creating a easier way to evaluate fees and a better user experience. 

It was mostly ignored and I did not get a lot of feedback on it unfortunatly. Its cool to see you inventing the same thing here

Anyway yes I think its a awesome idea - Current system suffer from problems that you correctly pointed out.

**Comment by vbuterin** [comment]
Paying full nodes is difficult because you can't identify who full nodes are reliably; it basically just degrades into some kind of proof of storage use.

> It is my understanding that quite often people spam transactions to create full blocks, especially mining cartels with a large portion of the hashrate, to force higher fees.

My approach would make spamming less effective in the short-term, though it would not change the effectiveness of a sustained long term attack. That's still a significant improvement though.

**Comment by prateektiwari7** [comment]
Hi @vbuterin 

I am reading the facts and algorithms from very small age. And I am not having such experience .
But @agaddy32 is some how right

As Caspher BFT and Instanbul BFT does works for you but for @agaddy32 query maybe it will not.

I totally agree with @vbuterin your formula but f[n+1] = f[n] * (1 + 0.25 * (w[n] / M - 0.5)) but what I suggest is a semaphore mechanism that will bind the node to particular transaction for f*d but with miner reward of 1/N .Due to majority of miners ledger Blockchain always have some faults .

1.Dependent ledger .I totally agree about miner POS block Production.But the ledger is not stable till majority of them are present , what would it be if majority miner left the ledger ?

2.Making the (virtual critical session) in POS by semaphore(encrpt(private,keyhold)) by miner in ledger will help POS to be better and stable even if majority of miners are not in it . The stabilty of formula will still remains f[n+1] = f[n] * (1 + 0.25 * (w[n] / M - 0.5)) .

**Comment by prateektiwari7** [comment]
[](https://github.com/EOSIO/eos/issues/4173)

Please check this EOS O(n2) peer faults caused

**Comment by prateektiwari7** [comment]
https://github.com/EOSIO/eos/issues/4173

Check this issue O(n2) taking to peers would never stable EOS in eco-cash environment.Every Peer have to talk twice in ledger .

**Comment by daira** [comment]
An important consideration here is that currently validation in blockchains is really expensive, because all full nodes replicate the same validation work which depends on the transaction bandwidth.

If we had a Coda-style succinct blockchain, then validation would be incredibly cheap, and in fact would not depend on the number of transactions (!). That is not to say that transactions don't still have overhead, but their remaining overhead would only be in mempool processing, not block validation.

I actually think it makes little sense to tweak around the edges of transaction fee economics when there are such massive gains to be made by known techniques, that would completely change the game, asymptotically.

**Comment by vbuterin** [comment]
There's a lot of problems with the "recursive ZKP will solve scalability" mentality:

1. It does not solve the data availability problem ( https://github.com/ethereum/research/wiki/A-note-on-data-availability-and-erasure-coding ), so truly validating the chain would still require downloading the data or at least the current state. So it's definitely NOT O(1). In fact, assuming ratio of bandwidth to computational power is constant, this single fact means that it's definitely not true that it changes anything at all "asymptotically" in the computer science definition of the term.
2. General-purpose succinct proof schemes have prover overhead of 100-100000x (and for realistic computation closer to 100000x), so in general you should not expect proofs to be ready in time for the next block. Hence, at the very least all currently online mining nodes, and nodes that care about fast confirmations, would still have to verify a block the traditional way. Realistically, proof generating nodes would be lagging behind the chain, so the proofs would primarily be useful for catching up with history (basically, it would give the same guarantees as fast syncing without the 51% assumption)
3. Even if only miners had to process transactions and everyone else could verify a magic proof, giving miners the ability to create arbitrarily large blocks is risky from a centralization perspective, because it would mean that becoming a profitable miner would have a high fixed cost in terms of bandwidth and hardware, making mining less accessible and less censorship-resistant.

So given that recursive ZKPs won't bring us into transaction post-scarcity land, fees will not go away, and so fee market optimizations are still valuable. And if they *did* take us into transaction post-scarcity land and made fees go away, the security of the blockchain would drop to zero once inflation stops in any case....

**Comment by daira** [comment]
I didn't say (or mean) that it would solve scalability. There will obviously be more work to do. But it will change things so much that it's not entirely clear what the remaining problems will be until it's implemented.

Coda is account-based, and you need an authentication path for the state of an account to make updates to it. But that is only held by the owner of the account. So there's nothing else to store globally. It actually does change the amount of data needed asymptotically, because you're storing the information needed to validate an account only once, not once per node.

It is not true that proofs would take longer to create than the block time. Verifying a pairing costs less than 8500 constraints (#3425), which is less than a tenth of the Sapling Spend circuit size. This isn't the full cost of validating (say) two Groth16 proofs, but a circuit to do that would still be much smaller than the Sapling Spend circuit.

We wouldn't allow arbitrary numbers of proofs in a block. There would still be a bandwidth limit, it would just be *much* larger. I don't have enough information to give a reliable estimate right now, but probably several orders of magnitude.

I'd be delighted if transaction fees stayed so low long-term as to cause a problem incentivising miners to mine! That would be a very good problem to have, and we can cross that bridge when we come to it.

**Comment by vbuterin** [comment]
> It is not true that proofs would take longer to create than the block time. Verifying a pairing costs less than 8500 constraints (#3425), which is less than a tenth of the Sapling Spend circuit size. This isn't the full cost of validating (say) two Groth16 proofs, but a circuit to do that would still be much smaller than the Sapling Spend circuit.

If you can make a proof for a zcash block within a few percent of the block time (it would have to be much less than the block time itself, otherwise the stale rate would go super-high if miners actually use it for validation), I will certainly be impressed; verifying a pairing in less than a hash is impressive already. Looking forward to seeing the attempt and learning from it however it goes!

**Comment by prateektiwari7** [comment]
The algorithm still need to be perfectly evaluated  

New EOS Bug Steals Resources Directly From Users

EOS, has made headlines in connection with a new issue. A new EOS bug has been discovered that allows stealing resources directly from a user. The resource in question is RAM, which is valuable in the EOS blockchain due to its scarcity.

The EOS blockchain aims to be a decentralized operating system and to allow the running of decentralized applications and smart contracts. Each decentralized application needs RAM to store data and the bigger the application, the more RAM it needs. However, until recently, RAM in the EOS blockchain has been limited to 64gb. EOS approved a solution that would increase the amount of RAM gradually. In the meantime, RAM has become a valuable resource in the ecosystem and users have been actively trading it. In this way, RAM has become a tradeable commodity with prices fluctuating based on demand and supply. According to the eos.feexplorer.io, 1kb of RAM currently costs 0.12 EOS, which translates into 60 cents

A malicious user can install code on their account which will allow them to insert rows in the name of another account sending them tokens. This lets them steal RAM by inserting large amounts of garbage into rows when dapps/users send them tokens.

**Comment by agaddy32** [comment]
> And if they did take us into transaction post-scarcity land and made fees go away, the security of the blockchain would drop to zero once inflation stops in any case....

Given that if mining reward went to approximately zero, and hashrate was dropping, it would be almost guaranteed that a hardfork or softfork activation rewriting the block reward would occur. Such a condition I think is an underthought and overtrusted expectation of the 21million effect. The hardcaps are not actually there, we only have distribution curves that are currently agreed upon.

Distribution curve is set until said curve no longer meets the needs of the network or its users. All of the current hardcaps will likely be removed when they get near, unless a massive fee market can be maintained, which will mean that the chain in question is not a very good means of exchange, which over time will reduce its value further reducing hashrate and easing the proposition for a new block reward algorithm.

Pretty much the only way that a hardcap will ever stand after the point which it comes into effect is if the network is too expensive to transact on, and such a network has surely failed its original purpose of being a peer to peer decentralized money.

I do understand that this is not a popular idea, nor should it be, but it is the truth. There are no hard caps, inflation will continue at agreed upon rates.

**Comment by daira** [comment]
I don't agree with automatically increasing the fee when blocks are full. That allows an attack by miners in which they escalate the standard fee by artificially "padding" the blocks they mine with dummy transactions. It's possible to construct padding transactions with very little computation and at zero monetary cost, and they need not be verified.

**Comment by daira** [comment]
Instead, I think that the standard fee should be kept roughly constant and small in terms of fiat (e.g. USD), as proposed in #2863. Scaling should be achieved by L1 and L2 protocol capacity increases &mdash; not by increasing fees, automatically or otherwise.

**Comment by vbuterin** [comment]
> I don't agree with automatically increasing the fee when blocks are full. That allows an attack by miners in which they escalate the standard fee by artificially "padding" the blocks they mine with dummy transactions. 

Padding blocks with dummy transactions is expensive in this scheme because the miner would need to pay the standard fee, half of which they lose. So it would actually be quite difficult for an attack to be profitable.

Also, pushing up transaction fees by restricting capacity is something that's possible for a majority of miners to do in any blockchain already; that's just a simple monopoly pricing problem, and the only way to avoid it is to make sure mining is sufficiently decentralized.

> ½ of the tx fee goes to the miner who mined that tx, the other ½ gets spread out over the next ~year's worth of miners.

It's important here that in the surge pricing mode, 95% should go to the miner and 5% get spread out. The reason is that with the scheme described as is, miners have the incentive to make private agreements with tx senders (eg. you could imagine big mining pools and exchanges doing this), where the tx sender pays the standard fee and then pays the miner 6.5x the standard fee out of band. So the tx sender pays 7.5x (instead of 10x) and the miner receives 7x (instead of 5x).

**Comment by zookozcash** [comment]
Related: http://www.dtc.umn.edu/~odlyzko/doc/paris.metro.pricing.pdf

**Comment by mtefagh** [comment]
Exciting work! I was wondering that in order to solve the first two problems about the volatility of transaction levels and inefficiencies of auctions, a more straightforward way in terms of Kolmogorov complexity as @vbuterin mentions in his paper might be to use the median of transaction fees instead of max or min. Both the problem of overpaying in the first price auction and the possibility of miners' manipulation in the second price auction reduce to the fact that in the first case max and the second case min functions are not stable and it is well-known that a much more robust statistic is median. You can make it even more robust if you take the median over a fixed number of previous blocks in the sense that even when a transaction was included in the last block we do not remove it for a limited time and it is still used to calculate the median. In this way, the transaction fees will become even predictable as they will be changing smoothly, and the transaction fees of consecutive blocks are approximately equal.

**Comment by mtefagh** [comment]
I have already mentioned the idea of a sliding median over multiple consecutive blocks in the previous comment. Additionally, one can also add a time lag as suggested by @AlexeyAkhunov in https://github.com/ethereum/EIPs/issues/1559#issuecomment-444845249

**Comment by mtefagh** [comment]
> it can do this for example by adjusting `f[n+1] = f[n] * (1 + 0.25 * (w[n] / M - 0.5))`,

Apart from my previous suggestion to use the median price auction either in the general setting or at least for the premiums in the @vbuterin's proposal, I also suggest considering an additive dynamics instead of the above multiplicative update rule. The proposed function is not jointly convex in `f[n]` and `w[n]` and this will make most of the current economic analyses not applicable to this framework. In general, most financial methods for many well-known tasks would become intractable. Specifically, I was studying this article on ["Optimal Execution of Portfolio Transactions"](https://pdfs.semanticscholar.org/3d2d/773983c5201b58586af463f045befae5bbf2.pdf) and the same transaction execution policies are immediately relevant to lower transaction fees and even more important than that, distribute the transactions across different blocks more evenly.

**Comment by mtefagh** [comment]
Another critical issue is that I believe @vbuterin should replace `w[n]` by `w[n+1]` in the above formula. Users are paying for the burden which the previous users put on the network rather than their own. As a result, if I have a batch of costly transactions and the network is underused, _i.e._ `w[n] < M/2`, I have incentive to let the transaction fee go down and collect my transactions to put them all in the first block that the network is overused,  _i.e._ `w[n] > M/2`. Generally, when the network is underused, you will refrain from executing transactions if they are not urgent which is an inefficient behavior.

**Comment by mtefagh** [comment]
Yet another point is that not only the previous attack reduces the total payment of the selfish users, but also it reduces the gas price for the whole network, and this one cannot be solved by substituting `w[n]` with `w[n+1]`. See [here](https://nbviewer.jupyter.org/github/mtefagh/fee/blob/master/fee.ipynb) for a simulation where only 10% of the transaction volume is enough to manipulate the network.

**Comment by vbuterin** [comment]
I think I see the flaw in your attack strategy. Your attack basically says that the attacker (with 10% of all transactions) concentrates their transactions in every tenth block, so nine blocks out of ten, the attacker's share is empty and the block size is 90% of the target, and the tenth time the attacker sends all their transactions so the block size is 190% of the target. This means that gas usage adjusts down by 1.25% nine blocks out of ten and adjusts up 11.25% the tenth time, but because of `(x-a)*(x+a) < x**2` inequalities, this leads to a slight decrease on average (`0.9875 ** 9 * 1.1125 ~= 0.9934` per 10-block period). Applying this 100 times, we get the observed 50% decrease. So it is true that it's possible to send the same transaction volume with slightly different consequences to demand, but this ignores the fact that demand adjusts with fees, and once the fee drops a few percent others will come in and the equilibrium re-established. All that this means is that in equilibrium average usage will be about 0.5% higher than the target; not a big deal.

That said, if we want to eliminate this issue entirely, we can do that by replacing `minfee *= (1 + excess * adjSpeed)` with `minfee *= e ** (excess * adjSpeed)`. Doing exponentiation in consensus is hard, but if we apply a quadratic Taylor approximation `e ** x = 1 + x + x**2 / 2`, so we do `minfee *= (1 + excess * adjSpeed + (excess * adjSpeed) ** 2 / 2)` , then the discrepancy drops from 1 - 0.9934 to 1 - 0.99978. I'm not sure if it's worth the extra complexity, I think either way works fine.

**Comment by vbuterin** [comment]
> Users are paying for the burden which the previous users put on the network rather than their own

This is intentional for incentive compatibility. It's a common technique in game theory for the price that a user pays to be explicitly only based on _other users'_ activity, so that incentive-compatibility for the user can be proven trivially as a simple matter of "if the price is favorable the user gets it, otherwise the user does not". Otherwise, a user's "real price" is higher than the "set price" because a user has to take into account the fact that their own transaction pushes up the real price for themselves.

**Comment by mtefagh** [comment]
> Otherwise, a user's "real price" is higher than the "set price" because a user has to take into account the fact that their own transaction pushes up the real price for themselves.

Yes, I agree with your observation. However, I think this can actually be a good point and should be exploited because it incentivizes a user to avoid executing a lot of transactions in a single block. This is similar to the problem of liquifying a large volume of some asset in game theory, in which whenever you sell some amount the price goes down, and as a result, you are incentivized to sell little by little. I think the same mechanism design helps to avoid network congestion.

To be more specific, I have already mentioned the paper ["Optimal Execution of Portfolio Transactions"](https://pdfs.semanticscholar.org/3d2d/773983c5201b58586af463f045befae5bbf2.pdf) in which the authors prove that if we have a very similar additive update formula, the optimal strategy is to execute transactions over a long period of time. Maybe we can use the same formula in here.

**Comment by mtefagh** [comment]
> That said, if we want to eliminate this issue entirely, we can do that by replacing `minfee *= (1 + excess * adjSpeed)` with `minfee *= e ** (excess * adjSpeed)`.

If I rewrite your formula as `log(minfee) += excess * adjSpeed` your alternative suggestion is just to have the same additive update rule and then exponentiate it to get `minfee`. As I said in the previous comment the additive framework is well-known in economics but why the exponential? In other words, suppose that `minfee += excess * adjSpeed`, and then we charge everyone a fee of `f(minfee)`. `f` could be the identity function, an affine mapping, or any other positive monotone function including the exponential function, but I am not sure why exponential seems like a canonical choice to you?

**Comment by mtefagh** [comment]
> This is intentional for incentive compatibility. It's a common technique in game theory for the price that a user pays to be explicitly only based on _other users'_ activity, so that incentive-compatibility for the user can be proven trivially as a simple matter of "if the price is favorable the user gets it, otherwise the user does not".

Even in the new settings corrected by the exponential or additive formula, I still believe that users should also be charged for their own transactions as well as the aggregated previous usage. Today, I was reading about the [gas token](https://gastoken.io/) and any such smart contract exploiting storage refund in Ethereum can be used for **hoarding** which is immediately profitable if the gas price you pay is independent of the gas cost of your smart contract.

**Comment by daira** [comment]
@vbuterin wrote:
> @daira wrote:
> > I don't agree with automatically increasing the fee when blocks are full. That allows an attack by miners in which they escalate the standard fee by artificially "padding" the blocks they mine with dummy transactions.
>
> Padding blocks with dummy transactions is expensive in this scheme because the miner would need to pay the standard fee, half of which they lose. So it would actually be quite difficult for an attack to be profitable.

That's not true. The adversary only needs one transaction to fill a block, so at the current standard fee rate they lose 0.00005 ZEC (~0.35 US cents) per block. It would clearly be economically rational for miners to form a coalition to escalate the fee to as much as the market will bear.

**Comment by vbuterin** [comment]
Let's look at a scenario more deeply. Suppose that the block size target is normalized to 1 (for mathematical convenience) and the current fee is 1. Suppose that the "natural" usage level is 1 at the current fee of 1 (so the system starts in equilibrium), and the demand is `D(fee) = 2 - fee`. Suppose the adjustment coefficient is as written above, so a full block adjusts the fee up by a factor of 9/8, and an empty block adjusts it down by a factor of 7/8.

An attacker fills a block, paying a fee of 1. In the next block, the fee is 9/8, so the demand is 7/8. In the block after that, the fee is 1 + 7/64, so the demand is 1 - 7/64. In the block after that the fee is 1 + 49/512, so the demand is 1 - 49/512.....

Altogether, the increase in fees paid plus deadweight loss is slightly less than 1/8 + 7/64 + 49/512..... = 1. (slightly less because of the users that drop out due to the increased fees). So the attacker burns 1 to gain a revenue of less than 1.

**Comment by daira** [comment]
That's an unrealistic demand function. If increasing fees by 1/9 results in a 1/8 reduction in demand, then they are already "as much as the market will bear". But fees *should* be *much* lower than this level. After all, fees are just a mechanism to avoid DoS (not necessarily a good one; just one that is simple to implement). We have no obligation to pay miners the maximum fee that the market will bear; that's what they get block rewards for.

**Comment by tromer** [comment]
Regarding whether the demand function in @vbuterin's example is realistic, note there are two parameter regimes:

1. If "total true demand" is higher than capacity, then for the fee market to modulate the demand, prices _must_ be close to "as much as the users will bear". This is the regime that @vbuterin considers.

2. If "total true demand" is lower than capacity then, as @daira says, we may strive to make fees low and consider them merely a way to disincentive DoS. This is the regime where Zcash is currently, and it's not clear if @vbuterin's example extends to it. 

Note that we can't technically distinguish "case 1" from "case 2 under DoS", so even if "total true demand" is low, under DoS we should expect prices to go up painfully: the legit users need to bid up the fees until the attacker is priced out of the market, and who folds first depends on the attackers' incentives. This seems inevitable.

**Comment by zookozcash** [comment]
Looks like the Ethereum devs are moving forward with applying a variant of this idea to Ethereum! https://github.com/ethereum/EIPs/issues/1559

**Comment by zookozcash** [comment]
Oh! I just stumbled upon Vitalik's paper that he presented at Crypto 2018 which explains the rationale for this design! https://ethresear.ch/t/draft-position-paper-on-resource-pricing/2838

**Comment by zookozcash** [comment]
Here's an excellent article by Haseeb Qureshi outlining the issues and three major proposed alternatives: https://hackernoon.com/blockchain-fees-are-broken-here-are-3-proposals-to-fix-them-1f772e1530dd

**Comment by daira** [comment]
Think carefully about the economics from the point of view of miners.

Because of difficulty adjustment, the marginal cost of mining adjusts to a point at which, roughly speaking, mining is only just worthwhile (since if it were more than only just worthwhile, more hash power would join and the difficulty would increase).

Now suppose that fees increase (for any reason; it doesn't matter why). The miners' average profit will increase. But this makes mining more attractive, so hash power increases, and so does difficulty. So the miners have approximately the same profit &mdash; but now they are *reliant* on the transaction fees. One might think that this is good (since it incentivises transaction inclusion), but wait. If the fees were already high, then now miners are reliant on the *even higher* fees. And so fees will escalate to whatever the market will bear, even though *neither users nor miners* are actually benefiting in the long term [*].

Contrary to naive intuition, the fees are *not* needed to pay for including transactions in blocks in any meaningful sense, but rather are at this level due to an artifact of having both a fee market and difficulty adjustment in the protocol. The transactions would have been included in blocks regardless, at a lower fee level.

When is it beneficial for miners to mine coinbase-only blocks anyway? Exactly when they would otherwise lose out on the block reward entirely, i.e. when they do not have time to fill in the block with transactions. The article acknowledges this. But the article fails to support its argument that this is a reason why we need a transaction fee market. On the contrary, if a miner doesn't have time to fill in a block without risking loss of the block reward, then it won't do so *regardless* of the transaction fee mechanism.

[*] Cui bono? Electricity companies, and to some extent ASIC manufacturers, but we shouldn't care about them.

**Comment by zookozcash** [comment]
Analysis of EIP-1559 (which is the Ethereum equivalent of this proposal here) and EIP-2593 (which is an alternative/complementary proposal).

https://github.com/ethereum/rig/blob/master/eip1559/notes-call3.md

https://insights.deribit.com/market-research/analysis-of-eip-2593-escalator/

**Comment by Flora103** [comment]
Hello beautiful Are you interested in having an honest sugar daddy and get allowance of $2000 who will spoil you financially and support you without requesting for anything sexual in return, feel free to dm me on my email:floralshrek@gmail.com
